torch/csrc/jit/README.md

```diff
@@ -1108,12 +1160,207 @@ TODO: differentiation, symbolic autograd,
+
+The table below shows the graph and code for this small `ScriptModule`:
+```python
+class M(torch.jit.ScriptModule):
+    @torch.jit.script_method
+    def forward(self, x, y, z):
+        # type: (Tensor, int, float) -> Tensor
+        if y > 2:
+            x = x + z
+        else:
+            x = x + y
+        return x
+
+m = M()
+```
+
+`m.graph`
+```
+graph(%x.1 : Tensor,
+      %y : int,
+      %z : float):
+  %5 : int = prim::Constant[value=1]()
+  %3 : int = prim::Constant[value=2]()
+  %4 : bool = aten::gt(%y, %3)
+  %x : Tensor = prim::If(%4)
+    block0():
+      %x.2 : Tensor = aten::add(%x.1, %z, %5)
+      -> (%x.2)
+    block1():
+      %x.3 : Tensor = aten::add(%x.1, %y, %5)
+      -> (%x.3)
+  return (%x)
+```
+
+`m.code`
+```python
+def forward(self,
+    x: Tensor,
+    y: int,
+    z: float) -> Tensor:
+  if torch.gt(y, 2):
+    x0 = torch.add(x, z, 1)
+  else:
+    x0 = torch.add(x, y, 1)
+  return x0
+```
+
+## Serialization
+
+[export.cpp](export.cpp)
+[pickler.cpp](pickler.cpp)
+[import.cpp](import.cpp)
+
+TorchScript programs are serialized with a call to `torch.jit.save()`. The resulting file (ending in `.pt` by convention) can be loaded/executed in C++ and Python.
+
+### Overview
+
+The `.pt` file is a zip archive (which can be opened with tools such as `unzip`) and contains:
+  * code - the Python printed graph of a module
+  * `model.json` - a JSON file of a model Protobuf def (defined in [torch.proto](caffe2/proto/torch.proto))
+  * `tensors/` - each of the tensors of the model, with their tensor storage stored directly in a file
+  * `attributes.pkl` - a Python `pickle` archive of the attributes of a module
+
+### `model.json`
+The `model.json` contains the structure information of the model. Each model must contain one main Module, and each module may contain multiple submodules, and each module contains a bunch of parameters (tensors). We serialize the metadata for each tensor inline in `model.json` (e.g., dims, strides, record name, etc).
+
+### `code/`
+
+The `code` directory contains the Python Printed `Graph`s of the main module and its submodules.
+
+### `tensors/`
+
+During export a list of all the tensors in a model is created. Tensors can come from either module parameters or Tensor type attributes. Metadata about each tensor is stored in `model.json` with an index into this list. The `data` field refers to the file which contains the tensor storage data. Tensors are saved by directly writing the Tensor storage to a file.
+
+`model.json`
+```json
+{
+  ...
+  "tensors": [
+    {
+      "dims": [
+        "40",
+        "800"
+      ],
+      "offset": "0",
+      "strides": [
+        "800",
+        "1"
+      ],
+      "requiresGrad": true,
+      "dataType": "FLOAT",
+      "data": {
+        "key": "tensors/0"
+      },
+      "device": "cpu"
+    }
+  ],
+  ...
+}
+```
+
+### `attributes.pkl`
```

Comment:
It would be nice to see an example pickle file, similar to how you have examples for model.json